### PR TITLE
Users

### PR DESCRIPTION
**Pull request by andarcabrera**
Hi,

Here is the pr for the following story:
- As a new user, I can create an account with a username and password
- Users should have username, email, password
- User email should be unique
- As an existing user, I can sign in with email address and password
- Items belong to a user
- A user has many listings
- A user has many purchases
- As an existing user I can log out

IMPORTANT - we are working with a live database. Please see instructions below:
1. rake db:migrate VERSION=20160525052400 (this will run the migrations up to and including 20160525052400)
2. rake update_data (this rake task will update the existing records in the db with the new fields)
3. rake db:migrate (this will run the rest of the migrations removing the unnecessary fields and adding constraints)

Note - the master is deployed to Heroku(https://powerful-spire-18110.herokuapp.com/) and running well with the exception of rendering the seed images, which we will try to look into.

Team Powerful Spire out.
Thanks,
AJ

### CODE REVIEW

File: app/controllers/purchases_controller.rb [hunk 1]
```diff
@@ -3,14 +3,17 @@
 
 class PurchasesController < ApplicationController
 
-  def new
-    item = Item.find(params[:item_id])
-    purchase = Purchase.new
-    @presenter = NewPurchasePresenter.new(purchase, item)
+  before_action :check_current_user
```
Comment (kevbuchanan) [inline app/controllers/purchases_controller.rb]: Seems like this should also be used in the Items controller for new and create.
- we added check_current_user to Application Controller and it's used both in purchases and items controller

File: app/controllers/purchases_controller.rb [hunk 2]
```diff
@@ -3,14 +3,17 @@
...
+
+  def check_current_user
```
Comment (kevbuchanan) [inline app/controllers/purchases_controller.rb]: This should be a private method. It may even need to go in ApplicationController or some other shared place dependig on the comment above.
- we moved it to application controller

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
+class SessionsController < ApplicationController
+  def new
```
Comment (kevbuchanan) [inline app/controllers/sessions_controller.rb]: If all this does is render the default sessions#new template, and this method is empty, then there's no reason to have this here.
- we deleted this

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  def password
+    @password ||= Password.new(password_hash)
+  end
+
+  def password=(new_password)
+    @password = Password.create(new_password)
+    self.password_hash = @password
+  end
+
+  def self.authenticate(session_params)
```
Comment (kevbuchanan) [inline app/models/user.rb]: Convention is to place class methods before instance methods.
- moved it above instance methods

File: lib/tasks/update_data.rake [hunk 5]
```diff
@@ -0,0 +1,25 @@
+task update_data: :environment do
+  items = Item.all
+  puts "Going to update #{items.count} items"
+
+  ActiveRecord::Base.transaction do
+    items.each do |item|
```
Comment (kevbuchanan) [inline lib/tasks/update_data.rake]: Imagine you had 1 million items in your database. This would load 1 million ActiveRecord objects in to memory and iterate through them, which probably would consume all of your memory. What you really want is something like [#find_each](http://api.rubyonrails.org/classes/ActiveRecord/Batches.html). Get rid of the call to .all and just use `Item.count` to get the count, then use #find_each to iterate through all the records.
- updated rake task

File: Gemfile [hunk 6]
```diff
@@ -14,6 +14,7 @@ gem 'sidekiq', '~> 4.1.2'
 gem 'faker'
 gem 'rails_12factor', group: :production
 gem 'redis'
+gem 'bcrypt'
```
Comment (alexcodreanu86) [inline Gemfile]: You should always lock down your versions of a gems.
- we locked down all gems
  QUESTION - isn't the Gemfile.lock's file role to lock the versions down? Do we need to lock down all gems or just gems that we add on top of what is generated with Rails new?

File: app/models/purchase.rb [hunk 7]
```diff
@@ -1,12 +1,14 @@
 class Purchase < ActiveRecord::Base
-  validates :email, :item_id, presence: true
+  validates:item_id, :purchaser_id, presence: true
```
Comment (alexcodreanu86) [inline app/models/purchase.rb]: There should either be a parenthesis or a white space between `validates:item_id`
- added white space

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    user = User.find_by(email: session_params[:email])
```
Comment (alexcodreanu86) [inline app/models/user.rb]: Is it important to know that these params come from the session? This is an http detail in the database layer
- renamed parameter to 'credentials'

File: app/views/items/index.html.erb [hunk 9]
```diff
@@ -2,7 +2,11 @@
   <div class="jumbotron, index-jumbotron">
     <div class="jumbotron-contents">
       <h1 class="text-center, text-in-jumbotron"> Items for Sale </h1>
-      <%= button_to "Add Item", items_new_path, method: :get, class: "btn btn-success, btn-add-item" %>
+      <% if current_user %>
+        <%= button_to "Add Item", items_new_path, method: :get, class: "btn btn-success, btn-add-item" %>
```
Comment (alexcodreanu86) [inline app/views/items/index.html.erb:6]: A small detail here, I don't think you need the node classes to be comma separated, white space is enough
- does not work without the comma. we had to keep it

File: app/views/layouts/application.html.erb [hunk 10]
```diff
@@ -1,15 +1,43 @@
...
+            </ul>
+          <% end %>
+          <ul class="nav navbar-nav">
+            <li class="active"> <%= link_to("Item Marketplace", items_path) %></li>
+          </ul>
+          <% if current_user %>
+            <ul class="nav navbar-nav navbar-right">
+              <li class="navbar-text">Hello <%= current_user.username.capitalize %></li>
+            </ul>
+          <% end %>
```
Comment (alexcodreanu86) [inline app/views/layouts/application.html.erb]: At the risk of having a few similarities I would prefer to have 2 partials, one for the logged in user and one for anonymous user.
- added partials

File: spec/controllers/users_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+      expect(User.last.username).to eq("pamplemousse")
+      expect(User.count).to eq(1)
+    end
+  end
+
+  context "new user is not created" do
+    it "response is successful" do
+      post :create, :user => {email: "lacroix@example.com", password: "chosenone" }
+
+      expect(response).to have_http_status(:success)
```
Comment (alexcodreanu86) [inline spec/controllers/users_controller_spec.rb]: should this return a success response when params are missing?
- returns 422 unprocessable_entity status

File: spec/controllers/sessions_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+  context "login is unsuccessful due to incorrect credentials" do
+    before(:each) do
+      User.create(username: "chosen", email: "thechosenone@example.com", password: "guitarman")
+    end
+
+    it "has a successful response" do
+      post :create, session: { email: "thechosenone@example.com", password: "ukaleileman" }
+
+      expect(response).to have_http_status(:success)
```
Comment (alexcodreanu86) [inline spec/controllers/sessions_controller_spec.rb]: Should this be a success response?
- same as above

File: spec/controllers/sessions_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+      expect(session[:user_id]).to be_nil
+    end
+  end
+
+  context "login is unsuccessful due to incomplete credentials" do
+    context "no email provided" do
+      it "renders login page" do
+        post :create, session: { password: "ukaleileman" }
+
+        expect(response).to have_http_status(:success)
```
Comment (alexcodreanu86) [inline spec/controllers/sessions_controller_spec.rb]: same as above
- same as above

File: spec/controllers/sessions_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+        expect(session[:user_id]).to be_nil
+      end
+    end
+
+    context "no password provided" do
+      it "renders login page" do
+        post :create, session: { email: "thechosenone@example.com" }
+
+        expect(response).to have_http_status(:success)
```
Comment (alexcodreanu86) [inline spec/controllers/sessions_controller_spec.rb]: See comments above
- 422

File: app/models/user.rb [hunk 15]
```diff
@@ -0,0 +1,26 @@
+require 'bcrypt'
+class User < ActiveRecord::Base
+  include BCrypt
+
+  validates :username, :email, :password_hash, presence: true
+  validates :email, uniqueness: true
+
+  has_many :listings, class_name: "Item", foreign_key: :seller_id
+  has_many :purchases, class_name: "Purchase", foreign_key: :purchaser_id
```
Comment (alexcodreanu86) [inline app/models/user.rb]: Are these associations used anywhere right now?
- we removed the purchases associations

File: app/models/item.rb [hunk 16]
```diff
@@ -1,6 +1,11 @@
 class Item < ActiveRecord::Base
-  validates :name, :description, :price, :email,  presence: true
-  scope :available, -> {  where(available: true) }
+  validates :name, :description, :price, :seller_id, presence: true
+
+  scope :available, -> { where(available: true) }
+
+  belongs_to :seller, class_name: "User", foreign_key: :seller_id
+  has_one :purchase
+  has_one :purchaser, through: :purchases, source: :purchaser
```
Comment (alexcodreanu86) [inline app/models/item.rb:7]: Are all these associations used?
- we removed the purchase associations. The other are used in the presenter

File: app/controllers/items_controller.rb [hunk 17]
```diff
@@ -11,9 +11,9 @@ def new
   end
 
   def create
-    @presenter = NewItemPresenter.new(Item.new(item_params))
+    @presenter = NewItemPresenter.new(current_user.listings.new(item_params))
```
Comment (alexcodreanu86) [inline app/controllers/items_controller.rb]: This is an unnecessary coupling to the User model association. The only thing that this is buying us is that it adds the seller_id to the model at the price of us relying on an association in the user model. I would recommend avoiding this type of couplings whenever possible.
- we made this more explicit and not dependent on listings

**Comment by alexcodreanu86** [comment]
About your 3 command migration: 
- imagine you have an automated deployment system setup to install dependencies, run your migrations and restart the app. Could you still deploy with the current setup?
- rake tasks should not be defined as scripts meant to be run just once. 
- Alex decided he likes our argument for this

File: app/controllers/items_controller.rb [hunk 18]
```diff
@@ -2,6 +2,16 @@
 require 'item_marketplace/presenters/new_item_presenter'
 
 class ItemsController < ApplicationController
+
+  before_action :check_current_user, only: [:new, :create]
+
+  def check_current_user
```
Comment (kevbuchanan) [inline app/controllers/items_controller.rb]: Why is this duplicated in both controllers?
- it's in application controller now

File: app/controllers/purchases_controller.rb [hunk 19]
```diff
@@ -20,7 +23,7 @@ def create
         redirect_to action: "show", item_id: item.id, id: purchase.id
       else
         @presenter = NewPurchasePresenter.new(purchase, item)
-        render :new
+        render "sessions/new"
```
Comment (kevbuchanan) [inline app/controllers/purchases_controller.rb]: Why does this render the new session template?
- it render items/index now

File: db/seeds.rb [hunk 20]
```diff
@@ -1,22 +1,16 @@
 require 'faker'
+require 'bcrypt'
```
Comment (kevbuchanan) [inline db/seeds.rb:2]: Why do you need to require and include Bcrypt here?
- removed the include but we need bcrypt to create the password hash

File: spec/controllers/items_controller_spec.rb [hunk 21]
```diff
@@ -34,46 +33,73 @@
   end
 
   context "new item is created successfully" do
+    before(:each) do
+      user =  User.create(username: "dreamteam", email: "us@dream.com", password_hash: "zzzzzzz")
```
Comment (kevbuchanan) [inline spec/controllers/items_controller_spec.rb]: 1) Pull this into a let
2) Move the line below to a method called login_user that does this. The name will explain what it is doing.
3) Call `login_user(user)` in your before block
4) Just use your `user` in your let anywhere you are doing `User.last`. Querying the last user is a less clear way of getting the specific user that you want to test with.
- we implemented this changes

File: spec/controllers/items_controller_spec.rb [hunk 22]
```diff
@@ -34,46 +33,73 @@
...
+      post :create, :item => {name: "", description: "A", price: 1, user_id: User.last.id}
 
       expect(Item.count).to eq(0)
     end
   end
-end
 
+  context "no current_user" do
+    before(:each) do
+      user =  User.create(username: "dreamteam", email: "us@dream.com", password_hash: "zzzzzzz")
```
Comment (kevbuchanan) [inline spec/controllers/items_controller_spec.rb]: Why do you assign to the user variable and not use it?
- we refactored this

Comment (kevbuchanan) [inline spec/controllers/items_controller_spec.rb]: Again, this could be a let. Your test setup should be clearer, as described in the above comment.
- in a let now

File: spec/controllers/purchases_controller_spec.rb [hunk 23]
```diff
@@ -2,71 +2,67 @@
...
-      expect(response.content_type).to eq("text/html")
-    end
-
-    it "renders the purchase template" do
-      get :new, :item_id => item.id
-
-      expect(response).to render_template(:new)
+  context "purchase successful" do
+    before(:each) do
+      session[:user_id] = purchaser.id
```
Comment (kevbuchanan) [inline spec/controllers/purchases_controller_spec.rb]: Again, this should be a method call. Put the method in a shared place.
- mase a login_user method

File: spec/controllers/sessions_controller_spec.rb [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+    it "redirects to the item index page" do
+      post :create, session: { email: "thechosenone@example.com", password: "guitarman" }
+
+      expect(response).to redirect_to items_path
+    end
+
+    it "creates a session for the user" do
+      post :create, session: { email: "thechosenone@example.com", password: "guitarman" }
+
+      expect(session[:user_id]).to eq(User.last.id)
```
Comment (kevbuchanan) [inline spec/controllers/sessions_controller_spec.rb]: Again, if you want to use the user later on, pull it into a let.
- moved to let

Comment (kevbuchanan) [inline spec/controllers/sessions_controller_spec.rb]: If you need the user to be created before calling `user`, you can use a `let!`, but some consider that poor form. The best way might be to just have a method:

``` ruby
user = create_user
# your test
```

File: app/models/user.rb (hunk 4 above)
Comment (kevbuchanan) [inline app/models/user.rb]: Also, is there a test for this? I don't see one in user_spec
- added tests

File: spec/controllers/users_controller_spec.rb [hunk 25]
```diff
@@ -39,13 +39,6 @@
   end
 
   context "new user is not created" do
-    it "response is successful" do
```
Comment (kevbuchanan) [inline spec/controllers/users_controller_spec.rb]: Why was this removed? Alex made a comment here asking whether the response should be successfull i.e. 200. He did not say to remove the test. There is something incorrect in your implementation that should be changed and reflected in this test.
- we added the proper response statuses

File: app/controllers/application_controller.rb [hunk 26]
```diff
@@ -1,3 +1,12 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+  include SessionHelper
+
+  def check_current_user(message)
+    if !current_user
+      flash[:notice] = message
+      render "sessions/new"
```
Comment (kevbuchanan) [inline app/controllers/application_controller.rb]: This should be a redirect.

-changed to redirect

File: spec/models/user_spec.rb [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+      expect(user.email).to eq("rogerthat@example.com")
+    end
+
+    it "has a password" do
+
+      expect(user.password).to eq("intriguing")
+    end
+
+    it "validates  the user" do
+      expect(User.authenticate(email: user.email, password: "intriguing")).to eq(user)
```
Comment (kevbuchanan) [inline spec/models/user_spec.rb:25]: This test setup is kind of confusing. I'd prefer to see a separate context for your authenticate method, and then one test that `it "returns the user for authentication with valid credentials"` and one that `it "returns nil for authentication with invalid credentials"`.
- refactored in separate context

File: Gemfile [hunk 28]
```diff
@@ -5,30 +5,34 @@ gem 'pg', '~> 0.15'
...
+gem 'jquery-rails', '~> 4.1.1'
+gem 'turbolinks', '~> 2.5.3'
 gem 'jbuilder', '~> 2.0'
 gem 'sdoc', '~> 0.4.0', group: :doc
 gem 'carrierwave', '~> 0.11.2'
 gem 'sidekiq', '~> 4.1.2'
-gem 'faker'
-gem 'rails_12factor', group: :production
-gem 'redis'
+gem 'faker', '~> 1.6.3'
```
Comment (kevbuchanan) [inline Gemfile:14]: This probably doesn't need to be a production dependency. Looks like you already have it listed below in development.
- we needed in production for heroku
